src/forum.cpp

```diff
@@ -1,147 +1,235 @@
-
-            if (vote_json.size() != 0) {
-                eosio_assert(vote_json[0] == '{', "vote_json must be a JSON object (if specified).");
-                eosio_assert(vote_json.size() < 4096, "vote_json should be shorter than 4096 bytes.");
-            }
-        }
-
-    private:
-
-        struct proposal {
-            name           proposal_name;
-            std::string    title;
-            std::string    proposal_json;
-
-            auto primary_key()const { return proposal_name.value; }
-        };
-        typedef eosio::multi_index<N(proposal),proposal> proposals;
-
-        struct statusrow {
-            account_name   account;
-            std::string    content;
-            time           updated_at;
-
-            auto primary_key() const { return account; }
-        };
-        typedef eosio::multi_index<N(status),statusrow> statuses;
-
-};
-
-EOSIO_ABI( forum, (post)(unpost)(propose)(unpropose)(vote)(status) )
+#include "forum.hpp"
+
+#define VALIDATE_JSON(Variable, MAX_SIZE)\
+validate_json(#Variable, Variable, MAX_SIZE)
+
+EOSIO_ABI(forum, (post)(unpost)(propose)(unpropose)(vote)(unvote)(status))
+
+// @abi
+void forum::post(
+    const account_name poster, 
+    const std::string& post_uuid, 
+    const std::string& content,
+    const account_name reply_to_poster, 
+    const std::string& reply_to_post_uuid, 
+    const bool certify,
+    const std::string& json_metadata
+) {
+    require_auth(poster);
+
+    eosio_assert(content.size() > 0, "content should be more than 0 characters long.");
+    eosio_assert(content.size() < 1024 * 10, "content should be less than 10 KB long.");
+
+    eosio_assert(post_uuid.size() > 0, "post_uuid should be longer than 3 characters.");
+    eosio_assert(post_uuid.size() < 128, "post_uuid should be shorter than 128 characters.");
+
+    if (reply_to_poster == 0) {
+        eosio_assert(reply_to_post_uuid.size() == 0, "If reply_to_poster is not set, reply_to_post_uuid should not be set.");
+    } else {
+        eosio_assert(is_account(reply_to_poster), "reply_to_poster must be a valid account.");
+        eosio_assert(reply_to_post_uuid.size() > 0, "reply_to_post_uuid should be longer than 3 characters.");
+        eosio_assert(reply_to_post_uuid.size() < 128, "reply_to_post_uuid should be shorter than 128 characters.");
+    }
+
+    VALIDATE_JSON(json_metadata, 8192);
+}
+
+// @abi
+void forum::unpost(const account_name poster, const std::string& post_uuid) {
+    require_auth(poster);
+
+    eosio_assert(post_uuid.size() > 0, "Post UUID should be longer than 0 characters.");
+    eosio_assert(post_uuid.size() < 128, "Post UUID should be shorter than 128 characters.");
+}
+
+// @abi
+void forum::propose(
+    const account_name proposer, 
+    const name proposal_name, 
+    const std::string& title,
+    const std::string& proposal_json
+) {
+    require_auth(proposer);
+
+    eosio_assert(proposal_name != N(proposal), "proposal name cannot be 'proposal'");
```

Comment:
now that we use the secondary index, we probably can take that out..

Reply:
True, there's no real "blocker" to have a proposal name called `proposal`

Removing this will reduce some valuable CPU time 😉 

```json
$ cleos get table eosforumdapp eoscanadacom proposal
{
  "rows": [{
      "proposal_name": "proposal",
      "title": "No issues with this",
      "proposal_json": ""
    }
  ],
  "more": false
}
```

Reply:
Yeah, this is a leftover from version 1 (multi table), will remove.